### PR TITLE
Handle Favorite

### PR DESCRIPTION
**Pull request by mvn-luongtran-dn**

### CODE REVIEW

**Comment by mvn-luongtran-dn** [comment]
Favorite Screen (Show and Delete Favorite)
Favorite Screen

Detail Screen (Add and Delete Favorite)
DetailScreen_1
DetailScreen_2

File: FinalProject/AppDelegate.swift [hunk 1]
```diff
@@ -30,7 +30,7 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
     func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
         window = UIWindow(frame: UIScreen.main.bounds)
         window?.makeKeyAndVisible()
-        Session.shared.token.isEmpty ? (setRoot(rootType: .login)) : (setRoot(rootType: .home))
+        (setRoot(rootType: .home))
```
Comment (mvn-quocdoan-dn) [inline FinalProject/AppDelegate.swift:33]: Reset please

File: FinalProject/Model/Core/APIManager.swift [hunk 2]
```diff
@@ -60,7 +60,7 @@ class ApiManager {
     func getDefaultHTTPHeaders() -> [String: String] {
         return [
             "Content-type": "application/json",
-            "Authorization": "Bearer \(Session.shared.token)"]
+            "Authorization": "Bearer eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJodHRwOi8vMTI3LjAuMC4xOjgwMDAvYXBpL3YxL3VzZXIvbG9naW4iLCJpYXQiOjE2Njc5Nzk0MzAsImV4cCI6MTY5OTUxNTQzMCwibmJmIjoxNjY3OTc5NDMwLCJqdGkiOiI1eDNGZVNVc2xCaXJ5Qmk3Iiwic3ViIjoiMjAiLCJwcnYiOiIyM2JkNWM4OTQ5ZjYwMGFkYjM5ZTcwMWM0MDA4NzJkYjdhNTk3NmY3In0.yLCAsKBkommq0KxhH8MfpJrVRMig6yUdVKEGWutGQn0"]
```
Comment (mvn-quocdoan-dn) [inline FinalProject/Model/Core/APIManager.swift:63]: Reset

File: FinalProject/Views/Controllers/Detail/DetailViewController.swift [hunk 3]
```diff
@@ -50,24 +52,35 @@ final class DetailViewController: UIViewController {
...
+        // Create bar button
         let backButton = UIBarButtonItem(image: UIImage(imageLiteralResourceName: "chevron"), style: .plain, target: self, action: #selector(returnButtonTouchUpInside))
         backButton.tintColor = .black
         navigationItem.leftBarButtonItem = backButton
 
-        let favoriteButton = UIBarButtonItem(image: UIImage(systemName: "heart.fill"), style: .plain, target: self, action: #selector(favoriteButtonTouchUpInside))
-        favoriteButton.tintColor = .red
+        favoriteButton = UIBarButtonItem(image: UIImage(systemName: "heart.fill"), style: .plain, target: self, action: #selector(favoriteButtonTouchUpInside))
+        // Check and update color favorite button
+        isFavorite = viewModel.isFavorite(product: product)
```
Comment (mvn-quocdoan-dn) [inline FinalProject/Views/Controllers/Detail/DetailViewController.swift:75]: No need this variable

File: FinalProject/Views/Controllers/Detail/DetailViewController.swift [hunk 4]
```diff
@@ -158,14 +179,13 @@ extension DetailViewController: UICollectionViewDataSource, UICollectionViewDele
...
             return UICollectionViewCell()
-
         }
         cell.viewModel = viewModel.viewModelForItem(at: indexPath)
         return cell
     }
 
     func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
-        return CGSize(width: collectionView.frame.width, height: collectionView.frame.height)
+        return CGSize(width: UIScreen.main.bounds.width, height: collectionView.frame.height - 5)
```
Comment (mvn-quocdoan-dn) [inline FinalProject/Views/Controllers/Detail/DetailViewController.swift:188]: 5, what does it mean?

File: FinalProject/Views/Controllers/Detail/DetailViewModel.swift [hunk 5]
```diff
@@ -6,19 +6,66 @@
...
-    #warning("Handle data product")
     var product: Product?
 
     init(product: Product?) {
         self.product = product
     }
 
+    var favoriteProducts: [Product] = []
     var currentIndex: Int = 0
+    var isFav: Bool = false
```
Comment (mvn-quocdoan-dn) [inline FinalProject/Views/Controllers/Detail/DetailViewModel.swift:21]: Remove please

File: FinalProject/Views/Controllers/Detail/DetailViewModel.swift [hunk 6]
```diff
@@ -6,19 +6,66 @@
...
+
+    func addFavoriteProduct() {
+        do {
+            let realm = try Realm()
+            try realm.write {
+                let product = Product(value: self.product ?? Product())
+                realm.add(product)
+            }
+        } catch {
+            fatalError()
```
Comment (mvn-quocdoan-dn) [inline FinalProject/Views/Controllers/Detail/DetailViewModel.swift:38]: Handle error please

File: FinalProject/Views/Controllers/Favorite/FavoriteViewController.swift [hunk 7]
```diff
@@ -78,11 +75,50 @@ extension FavoriteViewController: UITableViewDataSource, UITableViewDelegate {
...
+extension FavoriteViewController {
+
+    private func setUpObserve() {
+        do {
+            let realm = try Realm()
+            notificationToken = realm.objects(Product.self).observe({ [weak self] (_) in
+                guard let this = self else { return }
+                this.getFavoriteProduct()
+            })
+        } catch { }
```
Comment (mvn-quocdoan-dn) [inline FinalProject/Views/Controllers/Favorite/FavoriteViewController.swift:110]: Handle error

File: FinalProject/Views/Controllers/Search/SearchViewController.swift [hunk 8]
```diff
@@ -167,6 +167,6 @@ extension SearchViewController {
 
     func getProduct() {
         guard let viewModel = viewModel else { return }
-        viewModel.getProduct()
+//        viewModel.getProduct()
     }
```
Comment (mvn-quocdoan-dn) [inline FinalProject/Views/Controllers/Search/SearchViewController.swift:171]: You don't use this func

**Review by mvn-quocdoan-dn — APPROVED** [review]

**Pull request merged**